GUI System (see pr description)

123-WILL opened the pull request:
Can anchor UI Components to 9 normalized anchor positions
UI Component position offsets can be specified with either normalized or pixel coordinates
UI Component sizes can be specified with either normalized or pixel dimensions
Text Renderer integrated into TextBox Gui component
Text size can be specified in Point, Line Height (pixels), or Line Height (normalized)
Rectangle UI component for the color bar

123-WILL committed:
GUI Framework and Rectangle Rendering

123-WILL committed:
Added options for pixels vs normalized coordinates

123-WILL committed:
Fixed alpha blending

123-WILL committed:
GuiTextBox can now render text with a background color

123-WILL committed:
Commented out sample text box.

elsien-dev commented on src/engine/rendering/gui/GuiManager.cpp:
```diff
@@ -0,0 +1,59 @@
+
+namespace lei3d 
+{
+	GuiManager* GuiManager::s_instance = new GuiManager();
```
Does this get deleted?

123-WILL replied:
This should exist for the whole runtime of the app, I can't think of a scenario where it would need to be deleted.

elsien-dev commented on src/engine/rendering/gui/GuiManager.cpp:
```diff
@@ -0,0 +1,59 @@
+
+	void GuiManager::Init()
+	{
+		s_instance->m_guiFontShader = new Shader("./data/shaders/gui.vert", "./data/shaders/guifont.frag");
```
Same here. I'm pretty sure the default destructor won't delete these automatically.

123-WILL replied:
I don't think it will ever be deleted but I can add deleting it to the destructor anyway.

keptsecret replied:
Does it have to be allocated with `new`? These exist within the lifetime of the class singleton anyways.

kevinsadi commented:
before I review this, I apologize for the delay in reviewing / any merge conflicts you'll have to resolve before we merge this in.

elsien-dev commented on src/engine/rendering/gui/components/GuiComponent.hpp:
```diff
@@ -0,0 +1,68 @@
+			const std::pair<Space, glm::vec2>& pos = { NORMALIZED, { 0.25, 0.25 } }, 
+			const std::pair<Space, glm::vec2>& size = { NORMALIZED, { 0.5, 0.5 } }
+		);
+		virtual ~GuiComponent();
```
Good use of virtual destructor

123-WILL replied:
:D

elsien-dev reviewed (commented):
Can look at this a bit more later, but overall if the text rendering and other parts work. The only main issue I saw is that there are some instances where objects are created with new and not deleted.

keptsecret commented on src/engine/rendering/gui/GuiManager.cpp:
```diff
@@ -0,0 +1,59 @@
+	GuiManager& GuiManager::Instance()
+	{
+		return *s_instance;
+	}
```
Can you change this to the full C++ singleton pattern? I've changed existing ones recently, so refer to the SceneManager class for how to do it. Look at the constructor (make sure it's private), destructor and `GetInstance()` functions. This makes it thread-safe as well, among other things.

elsien-dev replied:
Just curious, what would make the original impl. unsafe? Is it bc s_instance can not be initialized? (I actually don't know when s_instance initializes since usually static vars are init at compile time, but can you actually initialize something at compile time using new? I'm guessing it's not lazy init.).

keptsecret replied:
Not gonna say I know for sure with this implementation but to be thread-safe, you'll need locks. The original implementation is doing something like [double-checked locking](https://en.wikipedia.org/wiki/Double-checked_locking). The change I made is the preferred way to do C++ singleton patterns since C++11, which is supported by modern compilers. The first time program flow enters the function the singleton instance is created and other concurrent flows will wait for it.

In this case, the memory is allocated at run-time rather than compile time and it also doesn't invoke the destructor on program termination because it was created  with `new`.

keptsecret commented on src/engine/rendering/gui/components/GuiComponent.hpp:
```diff
@@ -0,0 +1,68 @@
+		{
+			PIXELS,
+			NORMALIZED
+		};
```
I would switch these to `enum class` if you're not using the enumerator names for anything else. For regular enums, the values are in the same scope as the enum class itself, which can be confused with other enum values of the same name. Enum class values *have* to be called with the enum type like `Anchor::CENTER`, so this is just for additional safety.

You just have to change the type declarations here from `enum XXX` to `enum class XXX`.

keptsecret reviewed (changes requested):
Some changes to be made, largely on coding style choices to be (more) consistent with the rest of the project.

Also, resolve conflicts and it should be good to merge.

123-WILL committed:
Resolved merge conflict

123-WILL committed:
Revert "Resolved merge conflict"

This reverts commit e9e5f13a27f66fb2646a95e35ac0eec404af73f2.

123-WILL committed:
Merge branch 'main' into main

123-WILL committed:
Implemented changes requested in the PR

 - made singleton pattern consistent with the rest of the project
 - fixed a few possible memory leaks
 - made enum into enum class

elsien-dev reviewed (approved):
yay

kevinsadi commented on src/engine/rendering/gui/components/GuiComponent.hpp:
```diff
@@ -24,16 +25,16 @@ namespace lei3d
 		};
 
-		enum Space
+		enum class Space
```
Just for documentation and 🌟 learning 🌟  this is also referred to as "scoped enumeration."

kevinsadi replied:
It also not implicitly convertible to int like normal enums. We can specify an underlying type if we wanted one:

```
enum class Size : uint8_t {
    SMALL,
    MEDIUM,
    LARGE
};
```

123-WILL replied:
Ah thanks for explaining that, I was wondering why I had to cast it now. Interesting!

keptsecret reviewed (approved):
Looks great! Don't see any other problems with it. Should be good to merge.

kevinsadi commented:
lgtm

kevinsadi reviewed (approved):
the trifecta approval
